### PR TITLE
Slightly improve MinOpts JIT TP

### PR DESCRIPTION
**Pull request by EgorBo**
A few tricks to mitigate regressions which raised concerns in https://github.com/dotnet/runtime/pull/105190

PS: I still want to rewrite all OptimizationEnabled/Disabled into an OptLevel like Andy once suggested, but didn't want to do it here.

image

### CODE REVIEW

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/compiler.h [hunk 1]
```diff
@@ -10011,13 +10014,22 @@ class Compiler
...
+            assert(compMinOptsIsSet);
+            return !canUseAllOpts;
         }
         bool OptimizationEnabled() const
         {
-            return !OptimizationDisabled();
+            assert(compMinOptsIsSet);
+            return canUseAllOpts;
+        }
+        bool Tier0OptimizationEnabled() const
```
Comment (tannergooding) [inline src/coreclr/jit/compiler.h:10029]: A minor nit is that `OptimizationEnabled` and `Tier0OptimizationEnabled` both existing may be a bit confusing. Is there any idea what we would call the levels if this were some kind of `OptimizationLevel` enum, just for ideas on what we could call it as an alternative?

I think this is fine for now, because getting better names is a more involved change; just wanted to check if there were any other ideas/suggestions for the interim

File: src/coreclr/jit/gentree.cpp [hunk 2]
```diff
@@ -13340,10 +13507,7 @@ GenTree* Compiler::gtFoldExpr(GenTree* tree)
         return tree;
     }
 
-    // NOTE: MinOpts() is always true for Tier0 so we have to check explicit flags instead.
-    // To be fixed in https://github.com/dotnet/runtime/pull/77465
-    const bool tier0opts = !opts.compDbgCode && !opts.jitFlags->IsSet(JitFlags::JIT_FLAG_MIN_OPT);
-    if (!tier0opts)
+    if (!opts.Tier0OptimizationEnabled())
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.cpp:13510]: There's one more case in `gtFoldExprHWIntrinsic` that uses this pattern that should also be changed

File: src/coreclr/jit/vartype.h [hunk 3]
```diff
@@ -185,7 +185,7 @@ inline bool varTypeIsArithmetic(T vt)
 template <class T>
 inline bool varTypeIsGC(T vt)
 {
-    return ((varTypeClassification[TypeGet(vt)] & (VTF_GCR | VTF_BYR)) != 0);
+    return (TypeGet(vt) == TYP_REF) || (TypeGet(vt) == TYP_BYREF);
```
Comment (tannergooding) [inline src/coreclr/jit/vartype.h:188]: Was this an intentional change? Looks like it's a cheaper and better change, but also seems a bit unrelated to the changes above.

File: src/coreclr/jit/gentree.cpp [hunk 4]
```diff
@@ -6212,6 +6217,168 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
...
+    assert(tree);
+
+    if (tree->OperIsLeaf())
+    {
+        // Nothing to do for leaves.
+        return 0;
+    }
+
+    unsigned level = 1;
+    if (tree->OperIsSimple())
```
Comment (tannergooding) [inline src/coreclr/jit/gentree.cpp:6242]: nit: Might be worth a small comment saying that we are intentionally not handling hwintrinsics (they otherwise normally support GTF_REVERSE_OPS and other considerations).

**Review by tannergooding — APPROVED** [review]
This LGTM and is a very nice improvement overall.

Couple questions and one missed place where the new logic could also be used.

File: src/coreclr/jit/vartype.h (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/jit/vartype.h:188]: the intent of this PR was to slightly improve TP and this change seems to be doing so, we replace a memory load with two constants - it gives 0.1% improvement on all targets (although, a small regression on arm64 due to more instructions used for compare) as we eliminate the load

File: src/coreclr/jit/gentree.cpp (hunk 4 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:6242]: I'll add a comment, but we skip a lot of trees here, it was driven by asmdiff + tpdiff results

File: src/coreclr/jit/compiler.h (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/compiler.h:10029]: Yep, I agree that it doesn't look pretty, but a proper refactoring will be quite big and hard to review

Comment (EgorBo) [reply src/coreclr/jit/compiler.h:10029]: > any idea what we would call the levels if this were some kind of OptimizationLevel enum, just for ideas on what we could call it as an alternative?

Not yet. Andy suggested to use OptLevel: NoOpts, Size, Speed where Size implies "light-weight". The problem that we already have optLevel in jit, but we never use it (it's always Blended) so it all needs some re-thinking

File: src/coreclr/jit/gentree.cpp (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:13513]: Addressed

File: src/coreclr/jit/gentree.cpp (hunk 4 above)
Comment (tannergooding) [reply src/coreclr/jit/gentree.cpp:6247]: > but we skip a lot of trees here

👍, had mostly thought it might be worth a callout because it normally mirrors a lot of the `SimpleOp` handling (unlike any other node), so figured it might be worth a special note/callout

**Comment by EgorBo** [comment]
@jakobbotsch @tannergooding cc @dotnet/jit-contrib Could you please take another look as I've pushed more changes since your last approve.

TL;DR: I've introduced a stripped version of `gtSetEvalOrder` for MinOpts: `gtSetEvalOrderMinOpts` it doesn't calculate costs at all, so only operand swapping is preserved for certain operations. It contains a few quirks to minimize size regression impact.

As we discussed in Discord, it might makes sense to introduce a single IR walk e.g. before Rationalization to do the swap where profitable, but I am leaving this for a future PR.

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=752297&view=ms.vss-build-web.run-extensions-tab) with noticeable improvements for MinOpts as expected.

**Review by tannergooding — APPROVED** [review]

File: src/coreclr/jit/gentree.cpp [hunk 5]
```diff
@@ -6212,6 +6230,176 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
...
+//    the Sethi 'complexity' estimate for this tree (the higher
+//    the number, the higher is the tree's resources requirement)
+//
+unsigned Compiler::gtSetEvalOrderMinOpts(GenTree* tree)
+{
+    assert(tree);
+    if (fgOrder == FGOrderLinear)
+    {
+        // We don't re-order operands in LIR anyway.
+        return 0;
```
Comment (jakobbotsch) [inline src/coreclr/jit/gentree.cpp:6250]: Does it ever get called?

File: src/coreclr/jit/gentree.cpp [hunk 6]
```diff
@@ -6212,6 +6230,176 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
...
+        // Only GT_LEA may have a nullptr op1 and a non-nullptr op2
+        if (tree->OperIs(GT_LEA) && (op1 == nullptr))
+        {
+            std::swap(op1, op2);
+        }
+
+        // Check for a nilary operator
+        if (op1 == nullptr)
+        {
+            assert(op2 == nullptr);
```
Comment (jakobbotsch) [inline src/coreclr/jit/gentree.cpp:6274]: In what cases do we get here?

File: src/coreclr/jit/gentree.cpp (hunk 5 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:6250]: > Does it ever get called?

Not sure, but the original function `gtSetEvalOrder` contained this check

File: src/coreclr/jit/gentree.cpp [hunk 7]
```diff
@@ -6212,6 +6230,176 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
...
+        {
+            gtSetEvalOrderMinOpts(op1);
+            return level;
+        }
+
+        level             = gtSetEvalOrderMinOpts(op1);
+        unsigned levelOp2 = gtSetEvalOrderMinOpts(op2);
+
+        bool allowSwap = true;
+        // TODO: Introduce a function to check whether we can swap the order of its operands or not.
```
Comment (jakobbotsch) [inline src/coreclr/jit/gentree.cpp:6288]: Should this TODO be removed? Not sure I understand what it refers to. If it refers to extract the below logic into a function, why not just do that in this PR?

File: src/coreclr/jit/gentree.cpp (hunk 6 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:6274]: it's also copied from the original function. Looks like `GT_RETURN` for void is so.

File: src/coreclr/jit/gentree.cpp (hunk 7 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:6288]: I'd better remove the comment then for now, I don't mind. The idea is to share quirks with the Tier1 version (esp for STOREIND/STORE_BLK)

File: src/coreclr/jit/gentree.cpp [hunk 8]
```diff
@@ -6212,6 +6230,176 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
...
+                    tree->SetReverseOp();
+                    break;
+                }
+                if ((op1->gtFlags & GTF_ALL_EFFECT) != 0)
+                {
+                    break;
+                }
+
+                // In case op2 assigns to a local var that is used in op1, we have to evaluate op1 first.
+                if (gtMayHaveStoreInterference(op2, op1))
```
Comment (jakobbotsch) [inline src/coreclr/jit/gentree.cpp:6315]: You probably don't need this expensive version in MinOpts. It can be the conservative `(op2->gtFlags & GTF_ASG) != 0` that existed before #97409.

File: src/coreclr/jit/gentree.cpp (hunk 7 above)
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:6288]: Might be useful enough to leave a comment that it should be kept in sync with the quirks in the tier 1 version (and likewise in the tier 1 version)

File: src/coreclr/jit/gentree.cpp (hunk 8 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:6315]: > You probably don't need this expensive version in MinOpts. It can be the conservative `(op2->gtFlags & GTF_ASG) != 0` that existed before #97409.

just tried - it adds +28k bytes size regression for `benchmarks.run_pgo` collection

File: src/coreclr/jit/gentree.cpp (hunk 7 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:6299]: Moved both to a single helper, to keep the quirks in one place

File: src/coreclr/jit/gentree.cpp (hunk 8 above)
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp]: Ah ok, fine to leave it then. (Actually looks like I called out explicitly doing it in MinOpts in that PR as well)

**Review by jakobbotsch — APPROVED** [review]
